### PR TITLE
Working code ~library with hack...

### PR DESCRIPTION
**Pull request by Salamek**
Hi, this code works for me, it is kinda full library without all endpoints... you should split it into separate files (enums, exceptions etc) ...

There is one hack, and that is that ugly bunch of replace on line  cos Jackson XML serializer is ?broken? IDK... it is ignoring localName and uses variable name instead... you will need to report that to Jackson kotlin upstream and fix that or write some better looking workaround (like get those decorators from mutator and fix those tag names in doPostRequest or something)

### CODE REVIEW

**Comment by codpengu** [comment]
Damn awesome! Works so nice. Just had to add those variables to StateLogin. This xml parser isn't as good as the one u had tho :( for localNames you have to annotate a variable and a class with JacksonXmlProperty sadly. Extra work for the same result. I'll try to tinker it around and make it compatible for ios. I'll make sure to thank you in the developer notes man :D

**Comment by Salamek** [comment]
Can i see fixed working version with removed `replace` hack? 
I want to have some closure to this in my head :)

**Comment by codpengu** [comment]
Hmm im trying to figure that one out aswell now :D it was working before just fine

**Comment by codpengu** [comment]
Yeah i cant seem to fix it my self either, turns out JacksonXML is just a bad lib

**Comment by codpengu** [comment]
Found it! Pushed on ur PR. Going to merge it now :D

File: src/main/kotlin/com/qverkk/Main.kt [hunk 1]
```diff
@@ -335,16 +338,14 @@ class Huawei(url: String, username: String = "admin", password: String?) {
         return this.checkResponse(response, mutator)
     }
 
-    fun doPostRequest(endpoint: String, data: Any,  mutator: Class<*>): Any? {
+    fun doPostRequest(endpoint: String, data: Login, mutator: Class<*>): Any? {
```
Comment (Salamek) [inline src/main/kotlin/com/qverkk/Main.kt:341]: Note: Umm this method should be universal so data should be some parent of Login that is common for other data models... also variable loginRequestBody shoud be something more generic like requestBody...

File: src/main/kotlin/com/qverkk/Main.kt [hunk 2]
```diff
@@ -83,6 +85,8 @@ enum class LoginErrorEnum(val value: Int) {
 // DATA MODELS
 // ==============================================================================
 
+val xmlMapper = XmlMapper()
```
Comment (Salamek) [inline src/main/kotlin/com/qverkk/Main.kt:88]: Note: I wrote this kinda more object oriented... so this should be in Huawei class as property and called using this.xmlMapper ?

**Comment by codpengu** [comment]
Hey bro, could you take a look at why i'm getting this while trying to change CA Bands?
```
<?xml version="1.0" encoding="UTF-8"?>
<error>
<code>125003</code>
<message/>
</error>
```

Just make a new PR or something if you have time :D

File: src/main/kotlin/com/qverkk/Main.kt (hunk 1 above)
Comment (codpengu) [reply src/main/kotlin/com/qverkk/Main.kt:341]: It is set to Any in the newest version, i was tinkering around with it to see why it wasnt printing the xml values correctly :D

File: src/main/kotlin/com/qverkk/Main.kt (hunk 2 above)
Comment (codpengu) [reply src/main/kotlin/com/qverkk/Main.kt:88]: Oh yeah for sure, changing this

File: src/main/kotlin/com/qverkk/Main.kt (hunk 1 above)
Comment (Salamek) [reply src/main/kotlin/com/qverkk/Main.kt:341]: Ah ok

File: src/main/kotlin/com/qverkk/Main.kt (hunk 2 above)
Comment (codpengu) [reply src/main/kotlin/com/qverkk/Main.kt:88]: Updated in my repo ;)

**Comment by Salamek** [comment]
From Huawei LTE Routers SRC: var ERROR_WRONG_SESSION_TOKEN = 125003;
Aaand i dont have time to look on it right now...

**Comment by codpengu** [comment]
Thanks for the help bro :D also if you have any webiste with all the errors, i would be very greatful. And i think i need to refresh the sesInfo etc before executing that request i think

**Comment by Salamek** [comment]
```js
var ERROR_SYSTEM_NO_SUPPORT = 100002;
var ERROR_SYSTEM_NO_RIGHTS = "100003";
var ERROR_SYSTEM_BUSY = 100004;
var ERROR_LOGIN_USERNAME_WRONG = 108001;
var ERROR_LOGIN_PASSWORD_WRONG = 108002;
var ERROR_LOGIN_ALREADY_LOGIN = 108003;
var ERROR_LOGIN_USERNAME_PWD_WRONG = 108006;
var ERROR_LOGIN_USERNAME_PWD_ORERRUN = 108007;
var ERROR_LOGIN_USERNAME_PWD_MODIFY = 115002;
var ERROR_VOICE_BUSY = 120001;
var ERROR_WRONG_TOKEN = 125001;
var ERROR_WRONG_SESSION = 125002;
var ERROR_WRONG_SESSION_TOKEN = 125003;
```

**Pull request merged**